sql/analyzer/check_constraints.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"strings"
+
+	"github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/parse"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+// validateCreateCheck handles CreateCheck nodes, resolving references to "old" and "new" table references in
+// the check body. Also validates that these old and new references are being used appropriately -- they are only
+// valid for certain kinds of checks and certain statements.
+func validateCreateCheck(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scope) (sql.Node, error) {
+	ct, ok := node.(*plan.CreateCheck)
+	if !ok {
+		return node, nil
+	}
+
+	chAlterable, ok := ct.UnaryNode.Child.(sql.Table)
+	if !ok {
+		return node, nil
+	}
+
+	checkCols := make(map[string]bool)
+	for _, col := range chAlterable.Schema() {
+		checkCols[col.Name] = true
+	}
+
+	var err error
+	plan.InspectExpressionsWithNode(node, func(n sql.Node, e sql.Expression) bool {
+		if _, ok := n.(*plan.CreateCheck); !ok {
+			return true
+		}
+
+		// Make sure that all columns are valid, in the table, and there are no duplicates
+		switch expr := e.(type) {
+		case *expression.UnresolvedColumn:
```

Comment:
non-existent fields are necessarily unresolved here, added a case for `GetField` in case a different table's column resolves (not sure if that's valid)

Reply:
my bad deferred*, not unresolved